lib/spack/llnl/util/tty/log.py

```diff
@@ -811,16 +811,18 @@ def background_reader(reader, echo_writer, _kill):
                 # if echo: write line to user
                 while True:
                     is_killed = _kill.wait(.1)
+                    # Flush buffered build output to file
+                    # stdout/err fds refer to log file
                     self.stderr.flush()
                     self.stdout.flush()
+
                     line = reader.readline()
-                    while line:
-                        if self.echo:
-                            self.echo_writer.write('{0}'.format(line.decode()))
-                            self.echo_writer.flush()
-                        line = reader.readline()
+                    if self.echo and line:
+                        echo_writer.write('{0}'.format(line.decode()))
+                        echo_writer.flush()
 
```

Comment:
We should also `break`(and close the reader)  if `not line` (i.e. if the file is closed on the writer end).

Furthermore it occurs to me it might be worth enclosing this in a `try/finally` that closes the reader in the `finally` block (which would avoid failing to close the file in some exception cases).